R/lint.R

```diff
@@ -0,0 +1,279 @@
+    on.exit(setwd(owd))
+  }
+  options <- obj$options()
+  common <- list(default_namespace = obj$default_namespace(),
+                 language = language %||% obj$language(),
+                 prefix = options$interpolation[["prefix"]] %||% "{{",
+                 suffix = options$interpolation[["suffix"]] %||% "}}",
+                 escape = options$interpolation[["escape"]] %||% "- ")
+
+  usage <- lapply(expand_paths(path), lint_get_usage_file)
+  results <- lapply(usage, lint_compare_usage_file, obj, common)
+
+  attr(results, "title") <- title %||% "Translation lint report"
+  class(results) <- "lint_translations"
+  results
+}
+
+
+##' @export
+##' @rdname lint_translations
+lint_translations_package <- function(root, language = NULL) {
+  desc <- file.path(root, "DESCRIPTION")
+  if (!file.exists(desc)) {
+    stop(sprintf("Did not find DESCRIPTION at '%s'", root))
+  }
+  package <- read.dcf(desc, "Package")[[1]]
+  if (is.na(package)) {
+    stop("Invalid DESCRIPTION - failed to read package name")
+  }
+  ## TODO: See reside-90
+  loadNamespace(package)
+  obj <- translator(package = package)
+  title <- sprintf("{%s}", package)
+  lint_translations("R", obj, language, root, title = title)
+}
+
+
+##' Report on translation linting.  At present this produces a
+##' single-page html report that can be opened in a web browser.
+##' Future versions will include reporting to the terminal.
+##'
+##' @title Show translation lint report
+##'
+##' @param x A \code{lint_translations} object from
+##'   \code{\link{lint_translations}}
+##'
+##' @param file File to save the report in (default is to use a
+##'   temporary file).
+##'
+##' @return This function is called for the side effect of viewing a
+##'   report.  It will return (invisibly) the path to the produced
+##'   report.
+##'
+##' @export
+lint_translations_report <- function(x, file = tempfile()) {
+  html <- lint_translations_html_report(x)
+  browse_html(html, file = file)
+}
+
+
+lint_get_usage_file <- function(path) {
+  text <- readLines(path)
+  exprs <- parse(file = path, keep.source = TRUE)
+  data <- utils::getParseData(exprs)
+  data$depth <- cumsum((data$token == "'('") - (data$token == "')'"))
+  data$index <- seq_len(nrow(data))
+  idx <- data$token == "SYMBOL_FUNCTION_CALL" & data$text == "t_"
+  usage <- lapply(which(idx), lint_get_usage_expr, data)
+  list(path = path, text = text, data = data, usage = usage)
+}
+
+
+lint_get_usage_expr <- function(i, data) {
+  stopifnot(data$token[[i + 2L]] == "'('")
+  stopifnot(data$token[[i - 1L]] == "expr")
+
+  ## This is the overall expression id - if this is highlighted then
+  ## the whole relevant part of the translation will be.
+  ##
+  ## The more robust way of getting this looks to be to get the parent
+  ## of this, but we're working with basically undocumented data
+  ## structures here...
+  index <- data$index[[i - 1L]]
+
+  res <- parse_data_match_call(i, data, function(key, data, ...) {})
```

Comment:
I had some confusion with this. For me `res` and function name `parse_data_match_call` don't have much meaning. I wonder if a naming more in the domain of the problem we're trying to solve would be easier to read? Something like
`args <- get_function_args(index, data, order = c("key", "data", "..."))` would be clearer. Though definitely possible that just stems from unfamiliarity with `getParseData` and `match.call`. Or maybe some roxygen docs on `parse_data_match_call` itself about what the function trying to achieve would be useful.

Reply:
Expanded a bit in dababc7 and renamed `res` to `args` to be a bit clearer - does that help?